zebra-chain/src/sapling/tree.rs

```diff
@@ -108,71 +108,134 @@ impl From<&Root> for [u8; 32] {
+                        break;
+                    }
                 }
-                next_layer_up.push(merkle_crh_sapling(height, left, right));
             }
-
-            height += 1;
-            current_layer = next_layer_up.into();
         }
 
-        assert!(current_layer.len() == 1);
+        Ok(())
+    }
 
-        NoteCommitmentTree {
-            root: Root(current_layer.pop_front().unwrap()),
-            height,
-            count,
-        }
+    /// Returns the current root of the tree, used as an anchor in Sapling
+    /// shielded transactions.
+    pub fn root(&self) -> Root {
+        // Hash left and right together, filling in empty leaves as needed.
+        let leaf_root = merkle_crh_sapling(
+            0,
+            self.left.unwrap_or_else(|| EMPTY_ROOTS[0]),
+            self.right.unwrap_or_else(|| EMPTY_ROOTS[0]),
+        );
+
+        // Hash in parents up to the currently-filled depth, roots of the empty
+        // subtrees are used as needed.
+        let mid_root = self
+            .parents
+            .iter()
+            .enumerate()
+            .fold(leaf_root, |root, (i, p)| match p {
+                Some(node) => merkle_crh_sapling((i + 1) as u8, *node, root),
+                None => merkle_crh_sapling((i + 1) as u8, root, EMPTY_ROOTS[i + 1]),
+            });
+
+        // Hash in roots of the empty subtrees up to the final depth.
+        ((self.parents.len() + 1)..MERKLE_DEPTH)
+            .fold(mid_root, |root, d| {
+                merkle_crh_sapling(d as u8, root, EMPTY_ROOTS[d])
+            })
+            .into()
     }
-}
 
-impl NoteCommitmentTree {
     /// Get the Jubjub-based Pedersen hash of root node of this merkle tree of
-    /// commitment notes.
+    /// note commitments.
     pub fn hash(&self) -> [u8; 32] {
-        self.root.0
+        self.root().into()
+    }
+
+    /// Count of note commitments added to the tree.
+    ///
+    /// For Sapling, the tree is capped at 2^32.
+    pub fn count(&self) -> u32 {
+        self.parents.iter().enumerate().fold(
+            match (self.left, self.right) {
+                (None, None) => 0,
+                (Some(_), None) => 1,
+                (Some(_), Some(_)) => 2,
+                (None, Some(_)) => unreachable!(),
+            },
+            |acc, (i, p)| {
+                // Treat occupation of parents array as a binary number
+                // (right-shifted by 1)
+                acc + if p.is_some() { 1 << (i + 1) } else { 0 }
+            },
+        )
+    }
+
+    fn is_complete(&self) -> bool {
+        self.left.is_some()
+            && self.right.is_some()
+            && self.parents.len() == MERKLE_DEPTH - 1
+            && self.parents.iter().all(|p| p.is_some())
+    }
+}
+
+impl From<Vec<jubjub::Fq>> for NoteCommitmentTree {
+    /// Compute the tree from a whole bunch of note commitments at once.
+    fn from(values: Vec<jubjub::Fq>) -> Self {
```

Comment:
Code style: Would this be clearer as an `Extend` impl?
Do we want this conversion and also an `Extend` impl?

We should try to match whatever we've done with the history trees and Orchard.

Reply:
Same as above, it can return an error so we can't use Extend